apps/lockfile-explorer/src/commands/lint.ts

```diff
@@ -0,0 +1,167 @@
+      throw new Error(`ERROR: Detected inconsistent version numbers in package '${name}': '${version}'!`);
+    }
+
+    for (const [dependencyPackageName, dependencyPackageVersion] of Object.entries(
+      packages[dependencyPath].dependencies ?? {}
+    )) {
+      await checkVersionCompatibility(
+        shrinkwrapFileMajorVersion,
+        packages,
+        `/${dependencyPackageName}${shrinkwrapFileMajorVersion === 6 ? '@' : '/'}${dependencyPackageVersion}`,
+        requiredVersions,
+        checkedDependencyPaths
+      );
+    }
+  }
+}
+
+async function searchAndValidateDependencies(
+  rushConfiguration: RushConfiguration,
+  checkedProjects: Set<RushConfigurationProject>,
+  project: RushConfigurationProject,
+  requiredVersions: Record<string, string>
+): Promise<void> {
+  console.log(`Checking the project: ${project.packageName}.`);
+
+  const projectFolder: string | undefined = project.projectFolder;
+  const subspace: Subspace | undefined = project.subspace;
+  if (subspace && projectFolder) {
+    const shrinkwrapFilename: string = subspace.getCommittedShrinkwrapFilename();
+    const pnpmLockfileText: string = await FileSystem.readFileAsync(shrinkwrapFilename);
+    const doc = yaml.load(pnpmLockfileText) as Lockfile | LockfileV6;
+    const { importers, lockfileVersion, packages } = doc;
+    const shrinkwrapFileMajorVersion: number = getShrinkwrapFileMajorVersion(lockfileVersion);
+    const checkedDependencyPaths: Set<string> = new Set<string>();
+    for (const [relativePath, { dependencies }] of Object.entries(importers)) {
+      if (path.resolve(projectFolder, relativePath) === projectFolder) {
+        const dependenciesEntries = Object.entries(dependencies ?? {});
+        for (const [dependencyName, dependencyValue] of dependenciesEntries) {
+          const fullDependencyPath = `/${dependencyName}${shrinkwrapFileMajorVersion === 6 ? '@' : '/'}${
+            typeof dependencyValue === 'string'
+              ? dependencyValue
+              : (
+                  dependencyValue as {
+                    version: string;
+                    specifier: string;
+                  }
+                ).version
+          }`;
+          if (fullDependencyPath.includes('link:')) {
+            const dependencyProject: RushConfigurationProject | undefined =
+              rushConfiguration.getProjectByName(dependencyName);
+            if (dependencyProject && !checkedProjects.has(dependencyProject)) {
+              checkedProjects.add(project);
+              await searchAndValidateDependencies(
+                rushConfiguration,
+                checkedProjects,
+                dependencyProject,
+                requiredVersions
+              );
+            }
+          } else {
+            await checkVersionCompatibility(
+              shrinkwrapFileMajorVersion,
+              packages,
+              fullDependencyPath,
+              requiredVersions,
+              checkedDependencyPaths
+            );
```

Comment:
Seems like these results should be cached, as multiple projects can point at the same dependencies.

Reply:
The same dependency is never checked twice